src/main/java/org/spdx/utility/compare/LicenseCompareHelper.java

```diff
@@ -843,81 +844,173 @@ private static int findOriginalStart(int position, List<Pair<Integer, Integer>>
-			startIndex = findOriginalStart(matcher.start(), charPositions);
-			endIndex = findOriginalStart(matcher.end(), charPositions);
-			String completeText = normalizedText.substring(startIndex, endIndex);
-			try {
-				result = !isTextStandardLicense(license, completeText).isDifferenceFound();
-			} catch (SpdxCompareException e) {
-				logger.warn("Compare exception for license ID "+license.getLicenseId(), e);
-				result = false;
-			} catch (InvalidSPDXAnalysisException e) {
-				logger.warn("Compare exception for license ID "+license.getLicenseId(), e);
-				result = false;
+
+		return result;
+	}
+
+
+	/**
+	 * Returns a list of SPDX Standard License ID's from the provided list that were found in the text, using
+	 * the SPDX matching guidelines.
+	 * @param text Text to compare to the standard license texts
+	 * @param licenseIds License ids to compare against
+	 * @return List of SPDX standard license IDs from licenseIds that match
+	 * @throws InvalidSPDXAnalysisException If an error occurs accessing the standard licenses
+	 * @throws SpdxCompareException If an error occurs in the comparison
+	 */
+	public static List<String> matchingStandardLicenseIds(String text, List<String> licenseIds) throws InvalidSPDXAnalysisException, SpdxCompareException {
+		List<String> result = new ArrayList<>();
+
+		if (text != null && !text.isEmpty() && licenseIds != null && !licenseIds.isEmpty()) {
+			for (String stdLicId : licenseIds) {
+				SpdxListedLicense license = ListedLicenses.getListedLicenses().getListedLicenseById(stdLicId);
+				if (isStandardLicenseWithinText(text, license)) {
+					result.add(license.getLicenseId());
+				}
 			}
 		}
+
 		return result;
 	}
 
-	
 	/**
-	 * Returns a list of SPDX Standard License ID's that match the text provided using
+	 * Returns a list of SPDX Standard License ID's that were found in the text, using
 	 * the SPDX matching guidelines.
-	 * @param licenseText Text to compare to the standard license texts
-	 * @return Array of SPDX standard license IDs that match
+	 * @param text Text to compare to the standard license texts
+	 * @return List of SPDX standard license IDs that match
 	 * @throws InvalidSPDXAnalysisException If an error occurs accessing the standard licenses
 	 * @throws SpdxCompareException If an error occurs in the comparison
 	 */
-	public static String[] matchingStandardLicenseIds(String licenseText) throws InvalidSPDXAnalysisException, SpdxCompareException {
-		List<String> stdLicenseIds = ListedLicenses.getListedLicenses().getSpdxListedLicenseIds();
-		List<String> matchingIds  = new ArrayList<>();
-		for (String stdLicId : stdLicenseIds) {
-			SpdxListedLicense license = ListedLicenses.getListedLicenses().getListedLicenseById(stdLicId);
-			if (!isTextStandardLicense(license, licenseText).isDifferenceFound()) {
-				matchingIds.add(license.getLicenseId());
+	public static List<String> matchingStandardLicenseIds(String text) throws InvalidSPDXAnalysisException, SpdxCompareException {
```

Comment:
Since this is a public API, we should change the method signature (e.g. return value of an array to a list).  Even though I would like the change, it would probably break some of the library user's code.

Reply:
> Two high level design comments:
> 
> * the current implementation of `matchingStandardLicenseIds` (before this change) only matched the complete text and won't find a license text within any text before or after.

Note that issue #141 (which I understood that `matchingStandardLicenseIds` was intended to address) was written to provide exactly that.

By way of background this is my primary use case - to detect all of the potentially many licenses and exceptions contained in a single text (the [JavaMail license text](https://raw.githubusercontent.com/javaee/javamail/master/LICENSE.txt) being a good example of this in the wild).

> This was intentional due to the performance cost of finding licenses with the larger text. I would leave the current implementation for compatibility and performance reasons.  We could add a new method `matchingStandardLicenseIdsWithText` which would allow for non license text before and after.

Matching complete texts against specific licenses is already provided by `isTextStandardLicense`, so it seems duplicative to me to offer that functionality twice under different names?

> * There is a performance optimization I've been wanted to make for a while where we do a less precise but much faster match to narrow down a possible set of SPDX license ID's.  I haven't done this since it is a bit of work, but I'm thinking the optimization may be necessary for the `matchingStandardLicenseIdsWithinText`.

Absolutely.  I suspect finding the shortest unique substring in each template (e.g. via a generalised suffix tree) could be one good way to do this (especially as the set of unique substrings can be pre-generated per SPDX license list version).

That said, I'd propose getting the API and an initial (perhaps slow) implementation established first, even if it isn't as fast as it could be, then look at performance optimisation as a second step.

And FWIW the implementation of `matchingStandardLicenseIds` in this PR does not involve running the full license / exception regexes over the text in all cases; rather, it's a [two step process that involves determining whether a subset of the template is in the text (and if so extracting it), and then doing a full match against that subset](https://github.com/pmonks/Spdx-Java-Library/blob/issue-152/src/main/java/org/spdx/utility/compare/LicenseCompareHelper.java#L903-L906).

While it's definitely not the fastest potential design I can think of, it's also nowhere near as bad as just finding every full license / exception regex within the entire text.  I know this because that was the first naive approach I took, and it is _exceptionally_ slow - it can takes hours even on fairly simple texts (my suspicion is that some of the templates - the GPL ones in particular - result in regexes with catastrophic backtracking problems).

> The changes to include exceptions looks fine to me. I also don't think there is much of a performance problem if we're searching for a specific exception or license ID within the text - it is only a problem when we're looping over all the license/exception lists.

The latter being my use case: I'm confronted with a large set of unknown / 3rd party license texts with no hint or indication of which standard licenses (plural) and/or exceptions (plural) they might contain, and my code needs to figure that out, per text.  In fact that's exactly why the existing `isTextStandardLicense` method was of no use to me.

Reply:
> Note that issue https://github.com/spdx/Spdx-Java-Library/issues/141 (which I understood that matchingStandardLicenseIds was intended to address) was requesting exactly that.

> Matching complete texts against specific licenses is already provided by isTextStandardLicense, so it seems duplicative to me to offer that functionality twice under different names?

Not really duplicative since the functionality is slightly different and the performance is significantly different.  If we merge this PR, some of the existing code which calls this function will suffer a significant performance degradation.  To avoid this, we could create 2 functions with slightly different function (exact text vs. text with) and very different performance.

I suggest we go with the 2 methods approach.  I agree it is more complex, but I know if some implementations (e.g. the LicenseListPublisher) which will impact the larger community if we slow down the existing license.

Reply:
> That said, I'd propose getting the API and an initial (perhaps slow) implementation established first, even if it isn't as fast as it could be, then look at performance optimisation as a second step. 

Agree

Reply:
> While technically possible, it seems unlikely to me that any user code would be using these methods yet as they were only added in [v1.1.3](https://github.com/spdx/Spdx-Java-Library/releases/tag/v1.1.3), and 3 weeks isn't very much time for folks to have discovered the new methods, let alone started calling them from their own code.

This specific API has been in the code since the initial commit

Reply:
> This specific API has been in the code since the initial commit

Ah my misunderstanding - I had thought `matchingStandardLicenseIds` was new as of PR #152. I'll refactor all this out into actually new methods and revert all changes to `matchingStandardLicenseIds`.  Unless you suggest otherwise, I'll go with the names `matchingStandardLicenseIdsWithinText` and `matchingStandardLicenseExceptionIdsWithinText` for the new methods.

Reply:
Sound good - thanks1